digdag-standards/src/main/java/io/digdag/standards/operator/redshift/RedshiftConnection.java

```diff
@@ -197,26 +197,127 @@ String buildUnloadStatement(UnloadConfig unloadConfig)
-                            + "Or please ask system administrator to prepare a schema that\n"
-                            + "this operator can create tables in and specify `status_table` parameter"
-                            + "to create tables in the schema like `writable_schema.__digdag_status`",
-                            sql);
+            // do nothing
+        }
+
+        @Override
+        public boolean lockedTransaction(UUID queryId, TransactionAction action)
+                throws LockConflictException
+        {
+            beginTransaction();
+            boolean created = createLockedTableWithStatusRow(queryId);
+
+            if (created) {
+                // CREATE TABLE AS succeeded in this transaction.
+                // No other threads / nodes has succeeded to commit before.
+                // This transaction is responsible to run the action.
+                action.run();
+                updateStatusRowAndCommitTransaction(queryId);
+                return true;
+            }
+            else {
+                // CREATE TABLE AS conflicted.
+                // A thread / node succeeded to commit before.
+                // Skip the action.
+                abortTransaction();
+                return false;
+            }
+        }
+
+        private void beginTransaction()
+        {
+            executeStatement("begin a transaction", "BEGIN");
+        }
+
+        private void abortTransaction()
+        {
+            executeStatement("rollback a transaction", "ROLLBACK");
+        }
+
+        private void updateStatusRowAndCommitTransaction(UUID queryId)
+        {
+            executeStatement("update status row",
+                    String.format(ENGLISH,
+                        "UPDATE %s SET completed_at = CURRENT_TIMESTAMP WHERE query_id = '%s'",
+                        escapeIdent(statusTableName(queryId)),
+                        queryId.toString())
+                    );
+            executeStatement("commit updated status row", "COMMIT");
+        }
+
+        private void executeStatement(String desc, String sql)
+        {
+            try {
+                execute(sql);
+            }
+            catch (SQLException ex) {
+                throw new DatabaseException("Failed to " + desc, ex);
+            }
+        }
+
+        private boolean createLockedTableWithStatusRow(UUID queryId)
+        {
+            String sql = String.format(ENGLISH,
+                    "CREATE TABLE IF NOT EXISTS %s" +
+                    " (query_id, created_at, completed_at)" +
+                    " AS SELECT '%s'::text, CURRENT_TIMESTAMP::timestamptz, NULL::timestamptz",
+                    escapeIdent(statusTableName(queryId)),
+                    queryId.toString());
+            try {
+                execute(sql);
+                return true;
+            }
+            catch (SQLException ex) {
+                if (isConflictException(ex)) {
```

Comment:
T1
```
digdag_test=# begin; create table locktbl (i int);
BEGIN
CREATE TABLE
```
```
digdag_test=# select * from stv_locks;
 table_id |     last_commit     |        last_update         | lock_owner | lock_owner_pid |    lock_owner_start_ts     |  lock_owner_end_ts  |      lock_status
----------+---------------------+----------------------------+------------+----------------+----------------------------+---------------------+------------------------
   599721 | 2000-01-01 00:00:00 | 2017-01-04 02:51:36.01517  |    5941070 |           1597 | 2017-01-04 02:51:36.015169 | 2000-01-01 00:00:00 | Holding write lock
   599721 | 2000-01-01 00:00:00 | 2017-01-04 02:51:36.015173 |    5941070 |           1597 | 2017-01-04 02:51:36.015169 | 2000-01-01 00:00:00 | Holding  delete lock
   599721 | 2000-01-01 00:00:00 | 2017-01-04 02:51:36.015173 |    5941070 |           1597 | 2017-01-04 02:51:36.015169 | 2000-01-01 00:00:00 | Holding  insert lock
(3 rows)
```
T2
```
digdag_test=# begin; create table locktbl (i int);
BEGIN
(waiting....)
```
T1
```
digdag_test=# rollback;
ROLLBACK
```
T2
```
CREATE TABLE
digdag_test=#
```
T1
```
digdag_test=# begin; create table locktbl (i int);
BEGIN
(waiting...)
```
T2
```
digdag_test=# commit;
COMMIT
```
T1
```
ERROR:  duplicate key violates unique constraint "pg_class_relname_nsp_index" (possibly caused by concurrent transaction conflict)
digdag_test=#
```
`CREATE TABLE` with the same table name works as an exclusive lock 👍